### PR TITLE
*: support mysql compatible auto_increment, the client side changes

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### What is changed and how it works?

This is the client side changes.
It's split from https://github.com/pingcap/tidb/pull/38449 to make the life easier for the reviewers

### Check List

Tests 

- [X] Unit test

Test is in another PR https://github.com/pingcap/tidb/pull/38808

- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: autoid_service/autoid.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/pingcap/kvproto/pkg/autoid"
+	"github.com/pingcap/tidb/kv"
+	"google.golang.org/grpc"
+)
+
+type mockClient struct {
+}
+
+func (m *mockClient) AllocAutoID(ctx context.Context, in *autoid.AutoIDRequest, opts ...grpc.CallOption) (*autoid.AutoIDResponse, error) {
+	panic("TODO")
```
Comment (tiancaiamao) [inline autoid_service/autoid.go:15]: The mock one is here to pass build (after split PRs).
The functionality would be available after another PR been merged https://github.com/pingcap/tidb/pull/38810/files#diff-617839a099f419c84781f44a2286d763780a446a701eac0a440c7c579378706cR259

So the reviewers can skip here for now.

File: meta/autoid/autoid.go [hunk 2]
```diff
@@ -466,6 +468,7 @@ func (alloc *allocator) ForceRebase(requiredBase int64) error {
 	if requiredBase == -1 {
 		return ErrAutoincReadFailed.GenWithStack("Cannot force rebase the next global ID to '0'")
 	}
+
```
Comment (Defined2014) [inline meta/autoid/autoid.go:471]:
```suggestion
```

File: meta/autoid/autoid_service.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		DbID:       sp.dbID,
+		TblID:      sp.tblID,
+		N:          n,
+		Increment:  increment,
+		Offset:     offset,
+		IsUnsigned: sp.isUnsigned,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "rpc error") {
+			time.Sleep(200 * time.Millisecond)
```
Comment (Defined2014) [inline meta/autoid/autoid_service.go:130]: A little bit magic number, maybe use a constant.

File: meta/autoid/autoid_service.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	_, err = cli.Rebase(ctx, &autoid.RebaseRequest{
+		DbID:       sp.dbID,
+		TblID:      sp.tblID,
+		Base:       newBase,
+		Force:      force,
+		IsUnsigned: sp.isUnsigned,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "rpc error") {
+			time.Sleep(200 * time.Millisecond)
```
Comment (Defined2014) [inline meta/autoid/autoid_service.go:192]: ditto

File: meta/autoid/autoid_service.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+		defer span1.Finish()
+		ctx = opentracing.ContextWithSpan(ctx, span1)
+	}
+
+	if !validIncrementAndOffset(increment, offset) {
+		return 0, 0, errInvalidIncrementAndOffset.GenWithStackByArgs(increment, offset)
+	}
+
+retry:
+	cli, err := sp.GetClient(ctx)
```
Comment (Defined2014) [inline meta/autoid/autoid_service.go:114]: Why not save the `cli` in sp, reset it only when RPC meets error?

Comment (tiancaiamao) [reply meta/autoid/autoid_service.go:114]: Exactly as you said, see

https://github.com/pingcap/tidb/pull/38809/files#diff-f0f174acbfae2c5788672a929f39e218344beb43f732431b4fd542ce78b6a33bR63-R66

**Review by tangenta — APPROVED** [review]
